File: converter/src/test/kotlin/app/parsing/css/properties/longhands/effects/ClipPathPositionTest.kt
package app.parsing.css.properties.longhands.effects

// Regression suite for the wave-37 lane W3 clip-path repairs. Three
// separate defects, all of which silently produced a plausible-looking
// but wrong clip rather than failing loudly:
//
//   1. `parsePosition` understood only bare lengths plus a literal
//      `center`, so every keyword `at` clause in the corpus
//      (`circle(50% at left bottom)`, `circle(50% at right 40px bottom
//      40px)`, `ellipse(40px 60px at right top)`) resolved to null and
//      the shape re-centred. It now delegates to the shared
//      `PositionParser` (the css-values-4 <position> grammar extracted
//      from the wave-36 ObjectPositionPropertyParser rewrite).
//   2. The `rect()` branch of ClipPathShapeSerializer used
//      `value.top?.let { put("t", …) } ?: put("t", "auto")`. Because
//      `JsonObjectBuilder.put` returns the PREVIOUS element for the key —
//      always null on a fresh key — the elvis fired on the success path
//      too and every side was overwritten with "auto".
//   3. `shape()` (css-shapes-2 §4) had no parser branch at all, so the
//      declaration fell out of the IR entirely as an unmapped Generic.
//
// Pinned invariants below, one per behaviour, plus the byte-stability
// guarantee for the position forms that DID parse before this wave.

import app.irmodels.IRLength
import app.irmodels.properties.effects.ClipPathProperty
import app.irmodels.properties.effects.ClipPathShapeSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ClipPathPositionTest {

    private val json = Json

    /** Parse a bare `<basic-shape>` value down to its Shape. */
    private fun shapeOf(css: String): ClipPathProperty.Shape? {
        val prop = ClipPathPropertyParser.parse(css) ?: return null
        return when (val v = prop.value) {
            is ClipPathProperty.ClipPath.BasicShape -> v.shape
            is ClipPathProperty.ClipPath.GeometryBoxShape -> v.shape
            else -> null
        }
    }

    private fun circleOf(css: String) = shapeOf(css) as? ClipPathProperty.Shape.Circle
    private fun ellipseOf(css: String) = shapeOf(css) as? ClipPathProperty.Shape.Ellipse

    /** Assert one axis: value + unit + edge anchor. */
    private fun assertAxis(
        expectedValue: Double,
        expectedUnit: IRLength.LengthUnit,
        expectedEdge: String?,
        actualLength: IRLength,
        actualEdge: String?,
        label: String,
    ) {
        assertEquals(expectedValue, actualLength.originalValue, 1e-9, "$label value")
        assertEquals(expectedUnit, actualLength.originalUnit, "$label unit")
        assertEquals(expectedEdge, actualEdge, "$label edge")
    }

    // ── 1. the <position> grammar ────────────────────────────────────────

    @Test
    fun `two keywords resolve per axis, not positionally`() {
        // `left bottom`: x is the horizontal keyword, y the vertical one.
        // Normalised to the default origin — left is 0% from the left
        // edge, bottom is 100% from the top edge — so no edge anchor is
        // needed and consumers that ignore the edge field stay correct.
        val pos = assertNotNull(circleOf("circle(50% at left bottom)")?.position)
        assertAxis(0.0, IRLength.LengthUnit.PERCENT, null, pos.x, pos.xEdge, "x")
        assertAxis(100.0, IRLength.LengthUnit.PERCENT, null, pos.y, pos.yEdge, "y")
    }

    @Test
    fun `the unordered arm assigns by axis, so 'top right' is x=right`() {
        // The third arm of the grammar is an `&&`, i.e. unordered. A
        // positional reading would put `top` on x and swap the whole shape.
        val pos = assertNotNull(ellipseOf("ellipse(closest-side closest-side at top right)")?.position)
        assertAxis(100.0, IRLength.LengthUnit.PERCENT, null, pos.x, pos.xEdge, "x")
        assertAxis(0.0, IRLength.LengthUnit.PERCENT, null, pos.y, pos.yEdge, "y")
    }

    @Test
    fun `right and bottom offsets carry an edge anchor`() {
        // `right 40px` is 40px in from the RIGHT edge — not expressible as
        // a left-origin length without the reference-box width, which the
        // converter never sees.
        val pos = assertNotNull(circleOf("circle(50% at right 40px bottom 40px) border-box")?.position)
        assertAxis(40.0, IRLength.LengthUnit.PX, "right", pos.x, pos.xEdge, "x")
        assertAxis(40.0, IRLength.LengthUnit.PX, "bottom", pos.y, pos.yEdge, "y")
    }

    @Test
    fun `left and top offsets need no edge anchor`() {
        // `top 40px left 60px` — both offsets are already measured from
        // the default origin, and the unordered arm still has to route
        // them to the right axes.
        val pos = assertNotNull(
            ellipseOf("ellipse(farthest-side closest-side at top 40px left 60px) border-box")?.position
        )
        assertAxis(60.0, IRLength.LengthUnit.PX, null, pos.x, pos.xEdge, "x")
        assertAxis(40.0, IRLength.LengthUnit.PX, null, pos.y, pos.yEdge, "y")
    }

    @Test
    fun `a lone keyword centres the other axis`() {
        // Arm 1: `at left` is `left center`, NOT `left left`. The old
        // parser copied a single value onto both axes.
        val pos = assertNotNull(circleOf("circle(50% at left)")?.position)
        assertAxis(0.0, IRLength.LengthUnit.PERCENT, null, pos.x, pos.xEdge, "x")
        assertAxis(50.0, IRLength.LengthUnit.PERCENT, null, pos.y, pos.yEdge, "y")
    }

    @Test
    fun `'at center' keeps the exact pre-wave-37 wire shape`() {
        // The one position form the old parser did understand. Its bytes
        // must not move or every committed baseline shifts with it.
        val pos = assertNotNull(circleOf("circle(50% at center)")?.position)
        assertAxis(50.0, IRLength.LengthUnit.PERCENT, null, pos.x, pos.xEdge, "x")
        assertAxis(50.0, IRLength.LengthUnit.PERCENT, null, pos.y, pos.yEdge, "y")
        val obj = json.encodeToJsonElement(ClipPathShapeSerializer, circleOf("circle(50% at center)")!!).jsonObject
        // Edges default to null and `encodeDefaults` is off on the
        // converter's Json instance, so they must not appear on the wire.
        assertEquals(setOf("type", "r", "pos"), obj.keys)
        assertEquals(setOf("x", "y"), obj["pos"]!!.jsonObject.keys)
    }

    @Test
    fun `two bare percentages still parse as the ordered arm`() {
        val pos = assertNotNull(circleOf("circle(50% at 25% 75%)")?.position)
        assertAxis(25.0, IRLength.LengthUnit.PERCENT, null, pos.x, pos.xEdge, "x")
        assertAxis(75.0, IRLength.LengthUnit.PERCENT, null, pos.y, pos.yEdge, "y")
    }

    @Test
    fun `an invalid position drops the at-clause instead of guessing`() {
        // `top 10px` is not a position: `top` cannot name the horizontal
        // axis in the ordered two-component arm. A browser drops the whole
        // declaration; we drop the clause rather than emit a swapped guess.
        assertNull(circleOf("circle(50% at top 10px)")?.position)
        assertNull(circleOf("circle(50% at left right)")?.position)
    }

    // ── 2. the rect() elvis-on-put bug ───────────────────────────────────

    @Test
    fun `rect sides survive serialization`() {
        val shape = assertNotNull(shapeOf("rect(50px 200px 150px 50px)")) as ClipPathProperty.Shape.Rect
        assertEquals(50.0, shape.top?.originalValue)
        assertEquals(200.0, shape.right?.originalValue)
        val obj = json.encodeToJsonElement(ClipPathShapeSerializer, shape).jsonObject
        // The pre-fix wire was {"t":"auto","r":"auto","b":"auto","l":"auto"}
        // for EVERY rect() in the corpus.
        assertTrue(obj["t"]!!.jsonObject.containsKey("px"), "top must not be 'auto'")
        assertTrue(obj["l"]!!.jsonObject.containsKey("px"), "left must not be 'auto'")
        assertEquals(50.0, obj["t"]!!.jsonObject["px"]!!.jsonPrimitive.content.toDouble())
    }

    @Test
    fun `rect auto sides still serialise as the auto string`() {
        val shape = assertNotNull(shapeOf("rect(auto 200px auto 50px)")) as ClipPathProperty.Shape.Rect
        val obj = json.encodeToJsonElement(ClipPathShapeSerializer, shape).jsonObject
        assertEquals("auto", obj["t"]!!.jsonPrimitive.content)
        assertEquals("auto", obj["b"]!!.jsonPrimitive.content)
        assertTrue(obj["r"]!!.jsonObject.containsKey("px"))
    }

    // ── 3. shape() ───────────────────────────────────────────────────────

    @Test
    fun `shape() is preserved verbatim with whitespace collapsed`() {
        // WPT sources wrap these declarations across lines; the raw
        // newlines survive fixture extraction and must not reach the wire.
        val css = "shape(evenodd from 10px 10px,\n    hline by 80px, vline by 80%,\n    close)"
        val shape = assertNotNull(shapeOf(css)) as ClipPathProperty.Shape.ShapeFunction
        assertEquals(
            "shape(evenodd from 10px 10px, hline by 80px, vline by 80%, close)",
            shape.text,
        )
        val obj = json.encodeToJsonElement(ClipPathShapeSerializer, shape).jsonObject
        assertEquals("shape", obj["type"]!!.jsonPrimitive.content)
        assertEquals(shape.text, obj["fn"]!!.jsonPrimitive.content)
    }

    @Test
    fun `shape() keeps a trailing geometry box`() {
        val prop = assertNotNull(
            ClipPathPropertyParser.parse("shape(from center left, curve to center right with center top / center top, close) content-box")
        )
        val combined = prop.value as ClipPathProperty.ClipPath.GeometryBoxShape
        assertEquals("content-box", combined.box)
        assertTrue(combined.shape is ClipPathProperty.Shape.ShapeFunction)
    }

    // ── 4. path() fill-rule ──────────────────────────────────────────────

    @Test
    fun `path() splits its optional fill-rule off the path data`() {
        // Pre-fix, `d` came out as "nonzero, 'M0,0 L100,0 L0,100 L0,0'" —
        // fill rule and inner quotes baked into the path string.
        val shape = assertNotNull(shapeOf("path(nonzero, 'M0,0 L100,0  L0,100  L0,0')"))
            as ClipPathProperty.Shape.Path
        assertEquals("nonzero", shape.fillRule)
        assertEquals("M0,0 L100,0  L0,100  L0,0", shape.d)
    }

    @Test
    fun `path() without a fill-rule keeps its exact pre-fix wire shape`() {
        val shape = assertNotNull(shapeOf("path('M 0 0 L 10 10 Z')")) as ClipPathProperty.Shape.Path
        assertNull(shape.fillRule)
        assertEquals("M 0 0 L 10 10 Z", shape.d)
        val obj = json.encodeToJsonElement(ClipPathShapeSerializer, shape).jsonObject
        // `rule` must not appear — otherwise every existing path() fixture
        // shifts bytes.
        assertEquals(setOf("type", "d"), obj.keys)
    }

    @Test
    fun `a comma inside unquoted path data is not mistaken for a fill-rule`() {
        // Only `nonzero` / `evenodd` in the first slot are a fill rule; any
        // other head text is left attached so the quote strip can decide.
        val shape = assertNotNull(shapeOf("path('M0,0 L10,10')")) as ClipPathProperty.Shape.Path
        assertNull(shape.fillRule)
        assertEquals("M0,0 L10,10", shape.d)
    }

    @Test
    fun `an empty shape() is rejected`() {
        // Emitting an empty clip would hide the element entirely — worse
        // than dropping the declaration, which is what a browser does.
        assertNull(shapeOf("shape()"))
    }
}

File: runtimes/web/tests/effects/ClipPathPosition.test.ts
// ClipPathPosition.test.ts — pins the `at <position>` clause, the
// deep-flatten unwrapping and the shape()/rect() paths through the web
// extractor. Mirrors the Kotlin suite at
// src/test/kotlin/.../ClipPathPositionTest.kt so the invariant trips on
// both sides if either regresses.
//
// Wave-37 lane W3 bugs pinned here:
//  1. The Kotlin `parsePosition` accepted only bare lengths + a literal
//     `center`, so `circle(50% at left bottom)` reached the runtimes with
//     NO `at` clause at all. It now emits the full css-values-4 grammar,
//     with `xEdge`/`yEdge` for the right/bottom-anchored offset arm.
//  2. The IR property serializer deep-flattens a `{type, <one object
//     field>}` node, so `circle(50%)` arrives as
//     `{type:'circle', original:{…}}` and `circle(at 200px 150px)` as
//     `{type:'circle', x:…, y:…}`. Web handled neither: the first fell
//     through to the `closest-side` default (right only on a square box),
//     the second lost its centre.
//  3. `rect()` reached web as `rect(auto auto auto auto)` because of an
//     elvis-on-`put` bug in the Kotlin serializer; the round clause was
//     dropped here even when present.
//  4. `shape()` had no parser branch at all and fell out of the IR.

import { describe, it, expect } from 'vitest';
import { extractClipPath } from '../../src/engine/effects/clip/ClipPathExtractor';

const p = (type: string, data: unknown) => ({ type, data });
const css = (data: unknown) => extractClipPath([p('ClipPath', data)]).value;

// Percent / px IRLength wire shapes, spelled once.
const pct = (v: number) => ({ original: { v, u: 'PERCENT' } });
const px = (v: number) => ({ px: v });

describe('extractClipPath — at <position>', () => {

  it('keyword-only position arrives normalised to the default origin', () => {
    // `circle(50% at left bottom)` — the parser resolves the keywords to
    // 0% from left / 100% from top, so no edge field is needed and a
    // consumer that ignores edges is still correct.
    expect(css({ type: 'circle', r: pct(50), pos: { x: pct(0), y: pct(100) } }))
      .toBe('circle(50% at 0% 100%)');
  });

  it('right/bottom-anchored offsets keep the two-token CSS spelling', () => {
    // `circle(50% at right 40px bottom 40px)`. 40px in from the RIGHT
    // edge cannot be rewritten as a left-origin length without the box
    // width, so the IR carries the edge and we re-emit the author's
    // grammar — which CSS accepts verbatim inside `at`.
    expect(css({
      type: 'circle',
      r: pct(50),
      pos: { x: px(40), y: px(40), xEdge: 'right', yEdge: 'bottom' },
    })).toBe('circle(50% at right 40px bottom 40px)');
  });

  it('mixed anchoring: left-origin x, bottom-anchored y', () => {
    // `circle(50% at left 40px bottom 40px)` — only the y axis needs an
    // edge; x is already measured from the default origin.
    expect(css({
      type: 'circle',
      r: pct(50),
      pos: { x: px(40), y: px(40), yEdge: 'bottom' },
    })).toBe('circle(50% at 40px bottom 40px)');
  });

  it('ellipse carries the same at-clause treatment', () => {
    // `ellipse(farthest-side closest-side at top 40px right 60px)`:
    // keyword radii on both axes plus a right-anchored x.
    expect(css({
      type: 'ellipse',
      rx: 'farthest-side',
      ry: 'closest-side',
      pos: { x: px(60), y: px(40), xEdge: 'right' },
    })).toBe('ellipse(farthest-side closest-side at right 60px 40px)');
  });

  it('at center still serialises byte-identically to the pre-fix wire', () => {
    // The one position form the old parser DID understand. Its wire
    // shape must not shift, or every committed baseline moves.
    expect(css({ type: 'circle', r: pct(50), pos: { x: pct(50), y: pct(50) } }))
      .toBe('circle(50% at 50% 50%)');
  });
});

describe('extractClipPath — deep-flattened wire shapes', () => {

  it('flattened percentage radius is no longer read as closest-side', () => {
    // `circle(50%)` has ONE object-valued field, so the serializer inlines
    // the IRLength: {type:'circle', original:{v:50,u:'PERCENT'}}. Web had
    // a branch for the `px` spelling but not this one, and silently fell
    // back to `closest-side` — indistinguishable on a square box, wrong on
    // every other.
    expect(css({ type: 'circle', ...pct(50) })).toBe('circle(50%)');
  });

  it('flattened px radius keeps working (pre-existing branch)', () => {
    expect(css({ type: 'circle', ...px(40) })).toBe('circle(40px)');
  });

  it('flattened position (radius absent) is recovered', () => {
    // `circle(farthest-corner at 200px 150px)`: the corner keyword is not
    // in the css-shapes `<shape-radius>` grammar so the radius is dropped,
    // which leaves `pos` as the single object field and flattens it up.
    // The centre must still reach CSS.
    expect(css({ type: 'circle', x: px(200), y: px(150) }))
      .toBe('circle(closest-side at 200px 150px)');
  });

  it('a genuinely radius-less, position-less circle keeps the spec default', () => {
    expect(css({ type: 'circle' })).toBe('circle(closest-side)');
  });
});

describe('extractClipPath — rect() and shape()', () => {

  it('rect() emits its four sides, not four autos', () => {
    expect(css({ type: 'rect', t: px(50), r: px(200), b: px(150), l: px(50) }))
      .toBe('rect(50px 200px 150px 50px)');
  });

  it('rect() percentages survive', () => {
    expect(css({ type: 'rect', t: pct(25), r: pct(50), b: pct(75), l: pct(12.5) }))
      .toBe('rect(25% 50% 75% 12.5%)');
  });

  it('rect() auto sides still serialise as auto', () => {
    expect(css({ type: 'rect', t: 'auto', r: px(200), b: 'auto', l: px(50) }))
      .toBe('rect(auto 200px auto 50px)');
  });

  it('rect() round clause is emitted when the IR carries one', () => {
    expect(css({ type: 'rect', t: px(50), r: px(200), b: px(150), l: px(50), round: px(20) }))
      .toBe('rect(50px 200px 150px 50px round 20px)');
  });

  it('shape() passes through verbatim', () => {
    const fn = 'shape(evenodd from 10px 10px, hline by 80px, vline by 80%, close)';
    expect(css({ type: 'shape', fn })).toBe(fn);
  });

  it('shape() with a geometry box keeps the box suffix', () => {
    const fn = 'shape(from center left, curve to center right with center top / center top, close)';
    expect(css({ 'geometry-box': 'content-box', shape: { type: 'shape', fn } }))
      .toBe(`${fn} content-box`);
  });

  it('ellipse() with no radii defaults to closest-side, not a zero clip', () => {
    // WPT clip-path-ellipse-006 writes bare `ellipse()`. The absent axes
    // used to fall through to '0', i.e. `ellipse(0 0)` — an empty region
    // that erases the element.
    expect(css({ type: 'ellipse' })).toBe('ellipse(closest-side closest-side)');
  });

  it('path() keeps its fill-rule as a separate argument', () => {
    expect(css({ type: 'path', rule: 'nonzero', d: 'M0,0 L100,0 L0,100 L0,0' }))
      .toBe('path(nonzero, "M0,0 L100,0 L0,100 L0,0")');
  });

  it('path() without a fill-rule is unchanged', () => {
    expect(css({ type: 'path', d: 'M 0 0 L 10 10 Z' })).toBe('path("M 0 0 L 10 10 Z")');
  });

  it('an empty shape() payload is dropped rather than emitted', () => {
    // Defensive: an empty `fn` would produce the declaration `clip-path:`
    // which the browser discards anyway — better to emit nothing.
    expect(css({ type: 'shape', fn: '' })).toBeUndefined();
  });
});
